### ISSUE TITLE
String multi replacement with map for values

### ISSUE THREAD

**Issue by teddylear** [hashicorp/terraform#30260]
Ideally with the deprecation of the [template provider](https://registry.terraform.io/providers/hashicorp/template/latest/docs), there would a string function equivalent  for doing a multi replace in a string with a map given for values

### Use-cases
Use cases are for module deveoplers that sometimes needs to edit some user input and then inject that into other user input for complex queries. There are probably other use cases as well, but this is the first that comes to mind. 
### Attempted Solutions
There are some work arounds that people have done with current functions, but they become very complex very quick. One to note is this example from [this stackoverflow question](https://stackoverflow.com/questions/67375029/terraform-replacing-multiple-substrings-in-a-string). This seems like overkill for something fairly common. 

```terraform
locals {
  input = "Hello {name}, my favorite food is {food} I'd like to eat {food} and sleep."

  map = {
    food = "Sushi"
    name = "Data_sniffer"
  }

  out = join("\n", [
    for line in split("\n", local.input) :
      format(
        replace(line, "/{(${join("|", keys(local.map))})}/", "%s"),
        [
          for value in flatten(regexall("{(${join("|", keys(local.map))})}", line)) :
            lookup(local.map, value)
        ]...
      )
  ])
}

output "test_out" {
  value = local.out
}
```

### Proposal
Ideally there would be a function that takes a string then a map and does a replace based on what's in the map for values in {} if found.  For the above example there would be something like this: 
``` terraform
locals {
  input = "Hello {name}, my favorite food is {food} I'd like to eat {food} and sleep."

  map = {
    food = "Sushi"
    name = "Data_sniffer"
  }

  out = replaceValues(input, map)
}

output "test_out" {
  value = local.out
}
```
Please let me know if there are any questions/concerns regarding this. I'm open to any other implementation ideas, just want to make the end user experience easier for this common use case.

**Comment by crw** [comment]
Thanks for the enhancement request! I've added it to the list of things to triage.

**Comment by teddylear** [comment]
@crw If you want I can make a sample PR of what this would change would ideally look like to make things easier.

**Comment by crw** [comment]
Thanks for the offer. I think the description above does a pretty good job of describing what is desired, if a sample PR is desired I'll update this thread (or more likely, a dev will join the discussion to elaborate).

**Comment by jbardin** [comment]
See also #26838 for background

**Comment by teddylear** [comment]
@jbardin only issue with other proposal is that applies only to `templatefile` for input, where this proposal would solve problems with `templatefile` and building complex strings without `templatefile`.

**Comment by crw** [comment]
No, although thanks for the bump. This function would most likely need to be built as a plugin function provider, as opposed to a function built into Terraform. As such it is waiting for the implementation of plugin function providers. Please see:
* https://github.com/hashicorp/terraform/issues/2771
* https://github.com/hashicorp/terraform/issues/27696

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!

**Comment by Lillecarl** [comment]
This is implemented in https://registry.terraform.io/providers/random-things/string-functions/latest/docs/functions/multi_replace
